Check that a package contains a postinstall script before trying to run it

### Description of the bug

If postinstall is enabled for a package that doesn't have a postinstall script, you will see something like
```
npm ERR! Missing script: "postinstall"
npm ERR!
npm ERR! Did you mean one of these?
npm ERR!     npm install # Install a package
npm ERR!     npm uninstall # Remove a package
npm ERR!
npm ERR! To see a list of scripts, run:
npm ERR!   npm run
```


### Expected behavior

Flow should check if there is a postinstall script before trying to run it

### Minimal reproducible example

-

### Versions

- Vaadin / Flow version: 23 alpha 3

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
@@ -428,6 +428,19 @@ public class TaskRunNpmInstall implements FallibleCommand {
             if (!packageFolder.exists()) {
                 continue;
             }
+            try {
+                JsonObject packageJson = TaskGeneratePackageJson
+                        .getJsonFileContent(
+                                new File(packageFolder, "package.json"));
+                if (!containsPostinstallScript(packageJson)) {
+                    continue;
+                }
+            } catch (IOException ioe) {
+                logger.error(
+                        "Couldn't read package.json for {}. Skipping postinstall",
+                        ioe);
+                continue;
+            }
 
             logger.debug("Running postinstall for '{}'", postinstallPackage);
             try {
@@ -447,6 +460,11 @@ public class TaskRunNpmInstall implements FallibleCommand {
         lastInstallStats.installTimeMs = System.currentTimeMillis() - startTime;
         lastInstallStats.packageManager = enablePnpm ? "pnpm" : "npm";
 
+    }
+
+    private boolean containsPostinstallScript(JsonObject packageJson) {
+        return packageJson != null && packageJson.hasKey("scripts")
+                && packageJson.getObject("scripts").hasKey("postinstall");
     }
 
     private Process runNpmCommand(List<String> command, File workingDirectory)

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
@@ -433,6 +433,9 @@ public class TaskRunNpmInstall implements FallibleCommand {
                         .getJsonFileContent(
                                 new File(packageFolder, "package.json"));
                 if (!containsPostinstallScript(packageJson)) {
+                    logger.debug(
+                            "Skipping postinstall for '{}' as no postinstall script was found in the package.json",
+                            postinstallPackage);
                     continue;
                 }
             } catch (IOException ioe) {

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
@@ -56,6 +56,7 @@ import org.slf4j.LoggerFactory;
 import elemental.json.Json;
 import elemental.json.JsonObject;
 import net.jcip.annotations.NotThreadSafe;
+import org.slf4j.event.Level;
 
 @NotThreadSafe
 @Category(SlowTests.class)
@@ -353,6 +354,34 @@ public class TaskRunNpmInstallTest {
         FileUtils.write(packageJsonFile, packageJson.toJson(),
                 StandardCharsets.UTF_8);
 
+    }
+
+    @Test
+    public void runNpmInstall_noPostinstallScript_postIntstallNotExecuted()
+            throws IOException, ExecutionFailedException {
+        setupEsbuildAndFooInstallation();
+
+        // Remove postinstall script from "esbuild"
+        File esbuildPackageJson = new File(
+                new File(getNodeUpdater().nodeModulesFolder.getParentFile(),
+                        "fake-esbuild"),
+                "package.json");
+        JsonObject esbuildPackageJsonContents = Json.parse(IOUtils.toString(
+                getClass().getResourceAsStream(
+                        "fake-package-with-postinstall.json"),
+                StandardCharsets.UTF_8));
+        esbuildPackageJsonContents.getObject("scripts").remove("postinstall");
+        FileUtils.write(esbuildPackageJson, esbuildPackageJsonContents.toJson(),
+                StandardCharsets.UTF_8);
+
+        logger = new MockLogger();
+        Assert.assertTrue(logger.isDebugEnabled());
+        task.execute();
+
+        Assert.assertFalse(
+                "esbuild without postinstall should not have been executed",
+                ((MockLogger) logger).getLogs()
+                        .contains("Running postinstall for 'esbuild'"));
     }
 
     @Test

